tket/src/ZX/ZXDFromCircuit.cpp

```diff
@@ -0,0 +1,170 @@
+#include "Circuit/CircPool.hpp"
+#include "Transformations/Decomposition.hpp"
+#include "Transformations/Rebase.hpp"
+#include "Transformations/Transform.hpp"
+#include "ZX/ZXDiagram.hpp"
+
+namespace tket {
+
+namespace zx {
+
+enum class PortType { In, Out };
+typedef std::pair<VertPort, PortType> TypedVertPort;
+
+ZXDiagram::ZXDiagram(const Circuit& circ) : ZXDiagram() {
+  // Make a copy of the original circuit
+  Circuit c(circ);
+  // Rebase circuit to Rz, Rx, Hadamard, CZ, and CNOTs
+  OpTypeSet gates = {OpType::Rz, OpType::Rx, OpType::H, OpType::CX, OpType::CZ};
+  Transform t_decompose_box = Transforms::decomp_boxes();
+  Transform t_rebase =
+      Transforms::rebase_factory(gates, CircPool::CX(), CircPool::tk1_to_rzrx);
+  t_decompose_box.apply(c);
+  t_rebase.apply(c);
+  // Raise the scalar to the power of 2 due to doubling
+  multiply_scalar(exp(2. * i_ * PI * c.get_phase()));
+
+  sequenced_map_t<TypedVertPort, ZXVert> vert_lookup;
+  // Append each vertex to ZXDiagram raise error if not supported
+
+  BGL_FORALL_VERTICES(vert, c.dag, DAG) {
+    // We currently throw an error if the vertex is either
+    // 1. box , conditional, classical, flow, projective
+    Op_ptr op = c.get_Op_ptr_from_Vertex(vert);
+    if (is_box_type(op->get_type()) || is_flowop_type(op->get_type()) ||
+        is_classical_type(op->get_type()) ||
+        op->get_type() == OpType::Conditional) {
+      throw Unsupported(
+          "Cannot convert OpType: " + op->get_name() + " to a ZX node. \n");
+    }
+    switch (op->get_type()) {
+      case OpType::Input: {
+        ZXVert zx_vert = add_vertex(ZXType::Input, QuantumType::Quantum);
+        boundary.push_back(zx_vert);
+        vert_lookup.insert({{{vert, 0}, PortType::Out}, zx_vert});
+        break;
+      }
+      case OpType::Output: {
+        ZXVert zx_vert = add_vertex(ZXType::Output, QuantumType::Quantum);
+        boundary.push_back(zx_vert);
+        vert_lookup.insert({{{vert, 0}, PortType::In}, zx_vert});
+        break;
+      }
+      case OpType::ClInput: {
+        ZXVert zx_vert = add_vertex(ZXType::Input, QuantumType::Classical);
+        boundary.push_back(zx_vert);
+        vert_lookup.insert({{{vert, 0}, PortType::Out}, zx_vert});
+        break;
+      }
+      case OpType::ClOutput: {
+        ZXVert zx_vert = add_vertex(ZXType::Output, QuantumType::Classical);
+        boundary.push_back(zx_vert);
+        vert_lookup.insert({{{vert, 0}, PortType::In}, zx_vert});
+        break;
+      }
+      case OpType::Barrier:
+        continue;
```

Comment:
You might want to add a test for Barriers. I suspect that you will have an issue with edges since `vert_lookup` will not contain any information for the barrier vertex